### ISSUE TITLE
Add "before" and "after" hooks to ReadModels and, optionally, to Commands

### ISSUE THREAD

**Issue by alvaroloes** [boostercloud/booster#807]
## Description

Although the proposed solution for this issue covers several use cases, I'm going to explain the need for this with one use case that has more priority:

Right now, we have the possibility to authorize reading a ReadModel only to specific roles. This is great but not enough for any common application. Consider the following ReadModel:
```typescript
@ReadModel({
  authorize: [User]
})
class UserProfile {
   //...
}
```
Every client that is logged in with the role "User" can access **any instance of UserProfile**. This is normally incorrect and insecure. What we really want it to _allow accessesing the instance only if the current user is the owner of that profile_.

That is currently impossible in Booster, but needed for any application (for example Mayday would need this)

In Commands, this problem is not that important, because the user can check if the current user can access that command inside the command handler and throw an exception if they can't.
However, ReadModels doesn't have any place to add these kinds of checks, and that's what this this issue is for.

## Possible Solution
The proposed solution here adds a general mechanism that solves the above problem, but can also be used to other kind of use cases, like transforming data before sending it to the client, adding extra filters before executing the query, etc.

Proposal: Add two "hooks" to read models:

1. **before** hook: Executed before the query is made. It receives the `FilterFor` object with all the filter the user has specified and the current `UserEnvelope`. It can either throw exceptions or return a new `FilterFor` object to be used in the actual query sent to the provider. Very useful to reject requests under some conditions or further refine the filter before doing the actual query.
2. **after** hook: Executed after the query is made. It receives the array of ReadModels that are about to be sent to the user and the current `UserEnvelope`. It can throw exceptions or return the actual array of read models to be sent back to the user. Very useful for data transformations related to the presentational aspect.

Example:
```typescript
@ReadModel({
  authorize: [User]
})
class UserProfile {
  @Before // <--- This is a "before" hook
  public static checkID(filter: FilterFor<UserProfile>, currentUser?: UserEnvelope): FilterFor<UserProvile> {
    // Role "User" can only access their profile
    if (filter.id !== currentUser.id) {
      throw new NotAuthorizedError("...")
    }
    return filter
  }
  
  @After // <--- This is an "after" hook
  public static useProperCurrency(profiles: Array<UserProfile>, currentUser?: UserEnvelope): UserProfile {
    // We store all payments in cents of euros, but the user can choose the currency they want their money to be deployed in
    for(profile of profiles) {
      switch(profile.currency) {
        case 'dollars':
          profile.balance = toDollars(profile.balance)
          break
        case 'yen':
          profile.balance = toYens(profile.balance)
          break
      }
    }
    return profiles
  }
  //...
}
```

### Alternative sintaxes
The sintax proposed in the above example is one way of doing this. Let's call it way **A) Using decorators**. There are two extra alternative ways:

**Syntax B) Specify hooks in the ReadModel decorator**
```typescript
@ReadModel({
  authorize: [User]
  before: [UserProfile.checkID]
  after: [UserProvile.useProperCurrency]
})
class UserProfile {
  public static checkID(filter: FilterFor<UserProfile>, currentUser?: UserEnvelope): FilterFor<UserProvile> {...}
  public static useProperCurrency(profiles: Array<UserProfile>, currentUser?: UserEnvelope): UserProfile {...}

  // ...
}
```

**Syntax C) User a fixed name for hooks**
```typescript
@ReadModel({
  authorize: [User]
})
class UserProfile {
  public static before(filter: FilterFor<UserProfile>, currentUser?: UserEnvelope): FilterFor<UserProvile> {...}
  public static after(profiles: Array<UserProfile>, currentUser?: UserEnvelope): UserProfile {...}

  // ...
}
```

#### Discussion about syntaxes

Of all the syntax options, **C)** is the most restrictive, as only allow one hook per read model, while you might want to have several ones. Also the method names "before" and "after" doesn't show the business intention behind that code

Option **A)** seems pretty nice and allows you to both name your hooks as you want and have more than one "before" hook if you need to. The only drawback (if seen this way) is that all the hooks must be defined inside the ReadModel class, and you might have some generic hooks that would like to be reused in several read models.

Finally Option **B)** is the most flexible one: It has all the advantages of A) and it also allows you to set as hooks any function that might be defined anywhere, even inside a third party package. For example:

```typescript
import { toXML } from "xml-transformation-hooks"

@ReadModel({
  authorize: [User]
  before: [UserProfile.checkID]
  after: [UserProvile.useProperCurrency, toXML] // Now we return data in XML, for example
})
class UserProfile {
  public static checkID(filter: FilterFor<UserProfile>, currentUser?: UserEnvelope): FilterFor<UserProvile> {...}
  public static useProperCurrency(profiles: Array<UserProfile>, currentUser?: UserEnvelope): UserProfile {...}

  // ...
}
```

Another option is to **allow both option A and B**. This needs to be discussed.

## Aditional information
Adding this hooks to Commands would be a nice to have, but it is not a priority at all, because the `handle` method can do everything the hooks allow to do in ReadModels.

**Comment by javiertoledo** [comment]
It could make sense to have filters to do extra checks before letting the user access the data, but I'm not a super fan of having on-the-fly transformations in after filters, because optimizing data for reading is what read models are made for. For instance, in the example where you post-process the data to convert the currency, you could have several read models, one for each currency, or separate attributes for each currency that are updated on projection time and the user can select through GraphQL. I'm also worried that when people realize they can run code there and start doing weird things like querying other read models or implementing business logic just because some people could map that function with endpoint handlers they're used to from other frameworks.

**Comment by juanjoman** [comment]
I like the idea but I think it's more useful for other cases than for the one we're trying to solve. The cons I can think of (to specifically solve this problem) are:
* Users can do weirds things with logic in these hooks, not only solving the issue. We should aim to solve this concrete use case, without possible secondary effects.
* I don’t feel it’s a “booster-like solution“, since users have to apply custom logic on their own.
* Still, brute-forcing the id on queries is an issue

I'll leave below a couple of solution I've thought about:

### 1. Add a custom resource parameter in the user pool / auth provider. Example:
```json
'User #111': { 
  'role': 'User', 
  'resources': [      
    Cart: [<ids_array>],      
    Product: [<ids_array>]
  ] 
}
```
This way we could check if the resource id is inside the user’s resources. 

**Problem:** How can we dynamically store these params in the user pool? Also storing all new instances can be problematic with a high request volume

### 2. Add a restricted parameter on top of commands and read-models

Example in case that it’s not an annotation:

```typescript
@ReadModel/@Command({
  authorize: 'all',
  restricted: true // Optional, false by default
})
```
When the param is present, This will store an `ownedBy` parameter to entities when they’re store. This will be the id extracted from the Auth token (at command auth time). 

Then, from the read-model perspective, we can match the stored `ownedBy` id with the id taken from the Auth token again.

![image](https://user-images.githubusercontent.com/7570085/119667231-619b2100-be36-11eb-92e5-1172b62356d8.png)

I prefer the second option although it adds a bit of complexity to the commands registry and read-models queries' flow.

cc @javiertoledo @alvaroloes

**Comment by alvaroloes** [comment]
@javiertoledo Ok, we can start only with the `before` filter.

@juanjoman Take into account that users can do weird things anyways if they want. I understand that the brute-force ids can be a problem, but we need to take into account that the logic to check whether a user can access a specific read model is _business-dependent_. 
- What if the access to a read model doesn't depend on the user ID?
- What it depends on some other property in the auth token (because they are using a specific auth provider that adds more data to tokens)? 
- What if the read model belongs to many people and it contains an array with the IDs of people that can access it? We would need to check if the auth token ID is inside that array.
- Or what if the access to that read model depends on some field of the currently logged-in user that's not the ID?

This is the reason why that logic should be left to the developer to be written. We should help them up until the point where the business logic starts. That's what we have made with reducers, projections, etc. Those pieces of code can't be provided by Booster, because it can't possible know what they can be.
The same happens with this case.

Now focusing on your solution number 2, I didn't fully understand the role of the entities there:
1.- When a command is about to be executed and it is restricted: Which entity should we check for the `ownedBy` field?
2.- The same applies when we access to a read model: a read model is built based on many entities. Which one should we check?

**Comment by juanjoman** [comment]
Thanks for the explanation @alvaroloes! I just have a couple of questions:
1. Can `UserEnvelope` carry custom data? Isn't it just for `id`, `role` and `username`?

Ok if that's the case (having different ways of identify the "owner") then we should go with the @before hook proposal here.

About the solution number 2:
1. In the case that it exists, for that one. Otherwise none. This is because you could update an entity that it's "not yours"
2. Those that match the query + have the `ownerId` field

**Comment by juanjoman** [comment]
This was already merged on v0.17.0 (read-models) and v0.18.0 (commands)
